sqlserver/datadog_checks/sqlserver/connection.py

```diff
@@ -28,6 +28,81 @@ class SQLConnectionError(Exception):
     pass
 
 
+# we're not including space in the set of invalid characters because it's a valid key character for adodbapi
+# best be more permissive here so that the driver and the database have the final say as to whether it's valid or not
+CONNECTION_STRING_SPECIAL_CHARACTERS = set('=;[]{}"\'')
+
+
+def parse_connection_string_properties(cs):
+    """
+    Parses the properties portion of a SQL Server connection string (i.e. "key1=value1;key2=value2;...") into a map of
+    {key -> value}. The string must contain *properties only*, meaning the subprotocol, serverName, instanceName and
+    portNumber are not included in the string.
+    See https://docs.microsoft.com/en-us/sql/connect/jdbc/building-the-connection-url
+    """
+    cs = cs.strip()
+    params = {}
+    i = 0
+    escaping = False
+    key, parsed, key_done = "", "", False
+    while i < len(cs):
+        if escaping:
+            if cs[i : i + 2] == '}}':
+                parsed += '}'
+                i += 2
+                continue
+            if cs[i] == '}':
+                escaping = False
+                i += 1
+                continue
+            parsed += cs[i]
+            i += 1
+            continue
+        if cs[i] == '{':
+            escaping = True
+            i += 1
+            continue
+        # ignore leading whitespace, i.e. between two keys "A=B;  C=D"
+        if not key_done and not parsed and cs[i] == ' ':
+            i += 1
+            continue
+        if cs[i] == '=':
+            if key_done:
+                raise ConfigurationError(
+                    "Invalid connection string: unexpected '=' while parsing value at index={}: {}".format(i, cs)
+                )
```

Comment:
In this case it's supposed to raise a configuration error and stop the flow of the agent check if it fails to parse. We could have a custom exception but since this function is only for the SQL Server check I don't see it being used by other libraries. It's quite limited in the sort of parsing it does so I don't see it being used for much outside of this specific part of the check.